research/clustering/information_abstraction.py

```diff
@@ -0,0 +1,439 @@
+                else:
+                    if emd < min_emd:
+                        min_idx = idx
+                        min_emd = emd
+
+            # now increment the cluster to which it belongs -
+            turn_ehs_distribution[min_idx] += (
+                    1 / num_simulations
+            )
+        return turn_ehs_distribution
+
+    def get_river_ehs(self, num_print: int) -> np.ndarray:
+        """
+
+        :param num_print: number of simulations of opponents cards for calculating ehs
+        :return: np.ndarray of arrays containing [win_rate, loss_rate, tie_rate]
+        """
+        start = time.time()
+        river_ehs = [0] * len(self.river)
+
+        # iterate over possible boards/hole cards
+        for i, public in enumerate(self.river):
+
+            our_hand = list(public[:2])
+            board = list(public[2:7])
+
+            # get expected hand strength
+            game = GameUtility(our_hand=our_hand, board=board, cards=self._evals)
+            river_ehs[i] = self.simulate_get_ehs(game)
+
+            if i % num_print == 0:
+                print(
+                    f"Finding River Expected Hand Strength, iteration {i} of {len(self.river)}"
+                )
+        end = time.time()
+        print(f"Finding River Expected Hand Strength Took {end - start} Seconds")
+        return np.array(river_ehs)
+
+    def get_turn_ehs_distributions(self, num_print: int) -> np.ndarray:
+        """
+
+        :param num_print: frequency at which to print
+        :return: np.ndarray of distribution aware turn distributions
+        """
+        start = time.time()
+        turn_ehs_distributions = [0] * len(self.turn)
+
+        for i, public in enumerate(self.turn):
+            available_cards = [
+                x for x in self._evals if x not in public  # TODO need better implementation of this
+            ]
+
+            # sample river cards and run a simulation
+            turn_ehs_distribution = self.simulate_get_turn_ehs_distributions(
+                available_cards, the_board=list(public[2:6]), our_hand=list(public[:2]))
+
+            turn_ehs_distributions[i] = turn_ehs_distribution
+            if i % num_print == 0:
+                print(
+                    f"Finding Turn Distribution Aware Histograms, iteration {i} of {len(self.turn)}"
+                )
+        end = time.time()
+        print(f"Finding Turn Distribution Aware Histograms Took {end - start} Seconds")
+        return np.array(turn_ehs_distributions)
+
+    def get_flop_potential_aware_distributions(self, num_print: int, num_simulations: int = 5) -> np.ndarray:
+        """
+
+        :param num_print: frequency at which to print
+        :param num_simulations: number of simulations
+        :return: ndarray of potential aware histograms
+        """
+        start = time.time()
+        potential_aware_distribution_flops = [0] * len(self.flop)
+
+        for i, public in enumerate(self.flop):
```

Comment:
again we could make use of multi-threading if needs be lmk, would be good to also use tqdm if this is slow

Reply:
sounds good